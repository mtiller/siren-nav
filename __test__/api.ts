import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import { Siren } from "siren-types";

function s(obj: Siren): Siren {
    return obj;
}

export function usingMockAPI(tests: (mock: MockAdapter) => Promise<void>) {
    return async (): Promise<void> => {
        const mock = new MockAdapter(axios);

        //// Resources that deal with redirection and location headers
        mock.onGet("http://localhost/308").reply(308, "Wrong place, move along", {
            Location: "http://localhost/bar",
        });
        mock.onGet("http://localhost/resource").reply(
            200,
            {
                message: "You really should follow my location header",
            },
            {
                Location: "http://localhost/bar",
            },
        );
        mock.onGet("http://localhost/bar").reply(
            200,
            s({
                properties: {
                    message: "I am bar",
                },
            }),
        );
        mock.onGet("http://localhost/array").reply(200, new ArrayBuffer(10));
        mock.onGet("http://localhost/string").reply(200, "a string");
        mock.onGet("http://localhost/string").reply(200, 5);

        //// Resources that involve templates
        mock.onGet("http://localhost/search-template").reply(
            200,
            s({
                properties: undefined,
                links: [{ rel: ["search"], href: "/model/{model}" }],
            }),
        );

        //// Resources that provide collections
        mock.onGet("http://localhost/collection").reply(
            200,
            s({
                properties: undefined,
                links: [{ rel: ["item"], href: "/model/1" }, { rel: ["item"], href: "/model/2" }],
            }),
        );
        mock.onGet("http://localhost/model/1").reply(
            200,
            s({
                properties: { id: 1 },
                links: [{ rel: ["collection"], href: "/collection" }],
            }),
        );
        mock.onGet("http://localhost/model/2").reply(
            200,
            s({
                properties: { id: 2 },
                links: [{ rel: ["collection"], href: "/collection" }],
            }),
        );

        //// A self referential resource
        mock.onGet("http://localhost/self-ref").reply(
            200,
            s({
                properties: undefined,
                links: [{ rel: ["self"], href: "http://localhost/self-ref" }],
            }),
        );

        mock.onGet("http://localhost/api").reply(
            200,
            s({
                properties: undefined,
                actions: [
                    {
                        name: "query",
                        href: "/api/query",
                    },
                    {
                        name: "create",
                        href: "/api/create",
                        method: "POST",
                        type: "application/json",
                    },
                ],
            }),
        );
        mock.onGet("http://localhost/api/query").reply(
            200,
            s({
                properties: undefined,
                links: [],
            }),
        );
        mock.onPost("http://localhost/api/create").reply(
            200,
            s({
                properties: undefined,
            }),
        );

        try {
            await tests(mock);
        } catch (e) {
            throw e;
        } finally {
            mock.reset();
        }
    };
}
