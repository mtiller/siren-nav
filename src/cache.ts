import { Entity } from "siren-types";

export class Cache {
    private cache: { [key: string]: Entity<any> };
    constructor() {
        this.cache = {};
    }
    add(uri: string, resp: Entity<any>) {
        this.cache[uri] = resp;
    }
    has(uri: string): boolean {
        return this.cache.hasOwnProperty(uri);
    }
    get(uri: string): Entity<any> {
        if (this.has(uri)) return this.cache[uri];
        throw new Error("Requested uncached resource: " + uri);
    }
    getOr(uri: string): Entity<any> | null {
        if (this.has(uri)) return this.cache[uri];
        return null;
    }
}
