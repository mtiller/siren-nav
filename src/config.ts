import { AxiosRequestConfig } from "axios";

import * as debug from "debug";
const debugConfig = debug("siren-nav:config");

export type Config = AxiosRequestConfig;
export type Configurator = (cur: Config) => Config;

export const headerConfig = (key: string, value: string): Configurator => {
    return (config: Config) => {
        let newconfig = { ...config };
        if (!newconfig.headers) newconfig.headers = {};

        // NB - We must create a new header object!
        newconfig.headers = { ...config.headers };
        newconfig.headers[key] = value;

        debugConfig("  Set header '%s' to '%s'", key, newconfig.headers[key]);
        return newconfig;
    };
};

export const contentTypeConfig = (value: string): Configurator => {
    return headerConfig("Content-Type", value);
};

export const acceptConfig = (value: string): Configurator => {
    return headerConfig("Accept", value);
};
