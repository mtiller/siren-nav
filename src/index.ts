export * from "./navigation";
export * from "./requests";
export * from "./response";
export * from "./state";
export * from "./steps/apply";
export * from "./utils";
