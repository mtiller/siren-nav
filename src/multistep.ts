import { NavState } from "./state";
import { getSiren } from "./utils";
import { findPossible, Step } from "./steps";
import { Cache } from "./cache";

export type MultiStep = (cur: NavState, cache: Cache) => Promise<NavState[]>;

export function followEach(rel: string, parameters: {} | undefined): MultiStep {
    return (state: NavState, cache: Cache): Promise<NavState[]> => {
        return getSiren(state).then(siren => {
            const possible = findPossible(rel, siren, state, cache, parameters);
            return possible;
        });
    };
}

export function toMulti(step: Step): MultiStep {
    return async (cur: NavState, cache: Cache) => {
        const ns: NavState = await step(cur, cache);
        return [ns];
    };
}
