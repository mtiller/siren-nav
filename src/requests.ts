import { NavState } from "./state";
import { Siren } from "siren-types";
import axios from "axios";
import { AxiosResponse } from "axios";

import * as debug from "debug";
import { normalizeUrl, formulateData } from "./utils";
import { headerConfig } from "./config";
const debugRequests = debug("siren-nav:requests");

export interface ResponseData {
    data: any;
    headers: {};
    status: number;
}

export type Request = (cur: NavState) => Promise<ResponseData>;

export function performAction<T extends {}>(name: string, body: T, parameters?: {}): Request {
    return async (state: NavState): Promise<AxiosResponse> => {
        debugRequests("Performing action %s on %s", name, state.cur);
        debugRequests("  Fetching with current state of: %j", state);
        const resp = await getRequest(state);
        let siren: Siren = resp.data as Siren;
        // We may need to tweak the config, so let's start with the config of
        // the NavState
        let request = state.currentConfig;

        debugRequests("  Latest value of %s: %j", siren, state.cur);

        if (!siren.actions) {
            debugRequests("  ERROR: no actions defined in %s", state.cur);
            throw new Error("No actions defined for " + state.cur);
        }

        const action = siren.actions.find(action => action.name == name);
        if (!action) {
            throw new Error(
                `No action named ${name} found among: ${JSON.stringify(siren.actions.map(action => action.name))}`,
            );
        }
        debugRequests("  Found action with name %s", name);

        // Make sure there is an href here...
        if (!action.href) {
            debugRequests("    ERROR: no href");
            throw new Error("No href for action " + name);
        }

        // If the action specifies a content-type for the payload, include that
        // in the request.
        if (action.type) {
            debugRequests("    Content-Type set to %s", action.type);
            request = headerConfig("Content-Type", action.type)(request);
        }

        // Specify the method associated with the action (or fallback to "get")
        // Note that axios expects "get" or "post" (i.e., no GET, no POST).
        request = { ...request, method: (action.method || "get").toLowerCase() };

        let url = normalizeUrl(action.href, state.cur, parameters);

        let data: {} | string = formulateData(action, body);
        debugRequests("  Data for request: %j", data);

        // Add URL and data to request
        request = { ...request, url: url, data: data };

        debugRequests("    Making a request to %s with config %j", url, request);
        return Promise.resolve(axios.request(request));
    };
}

export const getRequest: Request = async (state: NavState): Promise<ResponseData> => {
    debugRequests("Requesting %s, current state is: %j", state.cur, state);
    debugRequests("  Fetching latest resource at %s with config %j", state.cur, state.currentConfig);
    return axios.get(state.cur, state.currentConfig).then(v => {
        debugRequests("  ResponseData for %s: %j", state.cur, v);
        return v;
    });
};
