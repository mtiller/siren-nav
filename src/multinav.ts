import { NavState } from "./state";
import { MultiStep, toMulti, followEach } from "./multistep";
import { Step, reduceEach, follow } from "./steps";
import { MultiResponse } from "./response";
import { getRequest } from "./requests";
import { Cache } from "./cache";

/**
 * The MultiNav class is not one a user would generally instantiate themselves.
 * Instead, it is created in response to certain steps in a chain that cause the
 * set of resources to "fan out".  The main example of this is when you follow a
 * relation and there are multiple instances of that relation.
 *
 * @export
 * @class MultiNav
 */
export class MultiNav {
    constructor(
        private start: Promise<NavState[]>,
        private steps: MultiStep[],
        private omni: Step[],
        private cache: Cache,
    ) {}
    get(): MultiResponse {
        let state = reduceEach(this.start, [...this.steps, ...this.omni.map(toMulti)], this.cache);
        let resp = state.then(s => Promise.all(s.map(x => getRequest(x))));
        return MultiResponse.create(resp);
    }
    do(step: Step): MultiNav {
        return new MultiNav(this.start, [...this.steps, toMulti(step)], this.omni, this.cache);
    }
    doMulti(steps: MultiStep): MultiNav {
        return new MultiNav(this.start, [...this.steps, steps], this.omni, this.cache);
    }

    follow(rel: string, parameters?: {}, which?: (states: NavState[]) => NavState): MultiNav {
        return this.do(follow(rel, parameters, which));
    }

    followEach(rel: string, parameters?: {}): MultiNav {
        return this.doMulti(followEach(rel, parameters));
    }
}
