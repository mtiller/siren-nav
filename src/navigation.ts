import { NavState, Config } from "./state";
import { Cache } from "./cache";
import { follow, Step, reduce, accept, auth, followLocation, header } from "./steps";
import { Entity } from "siren-types";
import { NavResponse } from "./response";
import { performAction, getRequest } from "./requests";
import { sirenContentType } from "siren-types";
import { Observable } from "rxjs";
import { MultiNav } from "./multi/multinav";

import * as URI from "urijs";
import { followEach, toMulti } from "./multi/multistep";

/**
 * The SirenNav class provides a collection of methods that allow for
 * a high-level declarative based approach to API navigation.
 *
 * @export
 * @class SirenNav
 */
export class SirenNav {
    /**
     * Create a new SirenNav instance for an API with an entry point
     * specified by the argument.
     *
     * @static
     * @param {string} url Initial URL
     * @param {string} base Base URL for API (all URIs will be anchored to this)
     * @returns
     *
     * @memberOf SirenNav
     */
    static create(url: string, cache?: Cache, config?: Config) {
        if (!cache) cache = new Cache();
        config = config || {};
        const uri = URI(url);
        if (uri.is("relative")) {
            throw new Error("SirenNav must be created with an absolute URL");
        }
        return new SirenNav(
            Promise.resolve(
                new NavState(
                    url,
                    undefined,
                    {
                        headers: {
                            Accept: sirenContentType, // Assume siren unless the user overrides it
                        },
                        // Commented out because it causes problems with HTTPS
                        // APIs that don't require authentication (for reasons
                        // I'm not 100% sure about).
                        // withCredentials: true,
                        ...config,
                    },
                    cache.getOr(url),
                ),
            ),
            [],
            [],
            cache,
        );
    }

    /**
     * Creates an instance of SirenNav.  This constructor is private because
     * it initializes lots of state that should be encapsulated (i.e., we
     * don't generally want people mucking about with this stuff).  Instead,
     * we use static functions to create instances.
     *
     * @param {Promise<NavState>} start
     * @param {Step[]} steps
     *
     * @memberOf SirenNav
     */
    private constructor(
        private start: Promise<NavState>,
        private steps: Step[],
        private omni: Step[],
        private cache: Cache,
    ) {}

    /**
     * Follow **one** instance of a given relation.  The optional second
     * argument is used to indicate that the first match should be used in cases
     * where multiple matches are available.  If the second argument is
     * undefined or false, then anything but a single (exact) match will result
     * in an error.
     *
     * If you wish to follow all links of a given relation, use followEach.
     *
     * @param {string} rel
     * @param {boolean} [first]
     * @returns {SirenNav}
     *
     * @memberOf SirenNav
     */
    follow(rel: string, parameters?: {}, which?: (states: NavState[]) => NavState): SirenNav {
        return this.do(follow(rel, parameters, which));
    }

    /**
     * Follow the location header in the response when fetching the current resource.
     *
     * @returns {SirenNav}
     * @memberof SirenNav
     */
    followLocation(): SirenNav {
        return this.do(followLocation);
    }

    followEach(rel: string, parameters?: {}): MultiNav {
        const multi = this.asMultiNav();
        return multi.doMulti(followEach(rel, parameters));
    }

    /**
     * Perform an action that expects a "hypermedia" payload as the input
     * to the function.
     *
     * @template P
     * @param {string} name
     * @param {Entity<P>} body
     * @returns {NavResponse}
     *
     * @memberOf SirenNav
     */
    performHyperAction<P extends {}>(name: string, body: Entity<P>): NavResponse {
        return this.performAction(name, body);
    }

    /**
     * Perform an action using an arbitrary type of input data.
     *
     * @template P
     * @param {string} name
     * @param {P} body
     * @returns {NavResponse}
     *
     * @memberOf SirenNav
     */
    performAction<P>(name: string, body: P): NavResponse {
        let state = reduce(this.start, [...this.steps, ...this.omni], this.cache);
        let resp = state.then(s => performAction(name, body)(s));
        return NavResponse.create(resp);
    }

    /**
     * Perform a navigation step.  This is mainly an internal method, but it
     * is exposed as a public method as to allow the set of possible
     * navigation steps to be extensible.  Any transformation satisfying
     * the Step type can be incorporated into the navigation process.
     *
     * @param {Step} step
     * @returns {SirenNav}
     *
     * @memberOf SirenNav
     */
    do(step: Step): SirenNav {
        return new SirenNav(this.start, [...this.steps, step], [...this.omni], this.cache);
    }

    /**
     * Perform a step that should *always* be performed.  These omni steps are done
     * after the navigation steps.  This is also an internal method, but it is made
     * public to allow extensions.
     */
    doOmni(step: Step): SirenNav {
        return new SirenNav(this.start, [...this.steps], [...this.omni, step], this.cache);
    }

    /**
     * This method can be used to generate a new SirenNav instance that
     * has already prenavigated a particular set of steps.  When a
     * "normal" SirenNav instance is evaluated, all the steps in the
     * navigation are repeated.  If the same navigation is repeated
     * over and over, many HTTP requests can potentially be made.
     * If those requests are sure to always return the same result,
     * then these requests will cause delays and unnecessarily
     * network requests.  The squash method performs all the
     * specified steps of the navigation BUT uses the resulting
     * state to create a new SirenNav instance that always starts
     * at the resulting state.  It therefore avoids repeating
     * steps on each execution.  However, squashing a SirenNav
     * is only appropriate if the preceding steps are sure to
     * always return the same results.
     *
     * @returns {SirenNav}
     *
     * @memberOf SirenNav
     */
    squash(): SirenNav {
        // NB - Note that we do NOT squash the omni steps.  Those remain.
        return new SirenNav(reduce(this.start, this.steps, this.cache), [], [...this.omni], this.cache);
    }

    /**
     * This tells the navigator to jump to a given URL.  This is
     * different from creating a new navigator because it
     * retains any information about the root URI of the API
     * and any configuration changes that have accumulated.
     *
     * @param {string} url
     * @returns {SirenNav}
     *
     * @memberOf SirenNav
     */
    goto(url: string, parameters?: {}): SirenNav {
        let newstate = new Promise<NavState>(async (resolve, reject) => {
            let state = await this.start;
            resolve(new NavState(state.rebase(url), parameters, state.config, this.cache.getOr(url)));
        });
        return new SirenNav(newstate, [], [...this.omni], this.cache);
    }

    /**
     * Add an entry to the Accept header.  This can be called multiple times and each
     * call will add an additional option.  As with the Accept header itself, order
     * matters.
     */
    accept(ctype: string): SirenNav {
        return this.do(accept(ctype));
    }

    /**
     * Authorize requests using the given scheme and token
     */
    auth(scheme: string, token: string): SirenNav {
        return this.do(auth(scheme, token));
    }

    /**
     * Add a given header
     *
     * @param {string} key The name of the header
     * @param {string} value The value to give to the header
     * @returns {SirenNav}
     * @memberof SirenNav
     */
    header(key: string, value: string): SirenNav {
        return this.do(header(key, value));
    }

    /**
     * This method returns the URL of the resource that is being pointed
     * to at the end of the chain of navigation steps.
     *
     * @returns
     *
     * @memberOf SirenNav
     */
    getURL(): Promise<string> {
        return reduce(this.start, [...this.steps, ...this.omni], this.cache).then(state => {
            if (state.config.baseURL) {
                return URI(state.cur)
                    .absoluteTo(state.config.baseURL)
                    .toString();
            }
            return state.cur;
        });
    }

    /**
     * The get method returns a NavResponse the represents the value of
     * the current resource.
     *
     * @returns {NavResponse}
     *
     * @memberOf SirenNav
     */
    get(): NavResponse {
        let state = reduce(this.start, [...this.steps, ...this.omni], this.cache);
        let resp = state.then(s => getRequest(s));
        return NavResponse.create(resp);
    }

    // TODO: Test this
    subscribe<T>(rel?: string, poll?: number): Observable<Entity<T>> {
        return new Observable(observer => {
            observer.complete();
            (async () => {
                // Follow rel
                // Get URL
                const url = await this.follow(rel || "events").getURL();
                if (url.startsWith("ws:" || url.startsWith("wss:"))) {
                    throw new Error("Unimplemented");
                    // Handle web socket case
                } else {
                    // Handle event source case
                    const source = new EventSource(url);
                    source.onmessage = msg => observer.next(msg as any);
                    // TODO: Distinguish between close and error
                    source.onerror = e => observer.error(e);
                }
            })();
        });
    }

    /**
     * Transforms the current SirenNav into a MultiNav (one capable of tracking
     * multiple concurrent resources)
     *
     * @protected
     * @returns {MultiNav}
     * @memberof SirenNav
     */
    protected asMultiNav(): MultiNav {
        const starts: Promise<NavState[]> = this.start.then(v => [v]);
        return new MultiNav(starts, this.steps.map(toMulti), this.omni, this.cache);
    }
}
