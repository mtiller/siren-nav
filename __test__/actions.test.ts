import { usingMockAPI } from "./api";
import { SirenNav } from "../src/navigation";

describe("Action tests", () => {
    it(
        "should perform a GET action without arguments",
        usingMockAPI(async mock => {
            const nav = SirenNav.create("http://localhost/api");
            const foo = nav.performAction("query", {});
            const r1 = await foo.asSiren();
            expect(r1).toMatchSnapshot();
        }),
    );
    it(
        "should perform a GET action with arguments",
        usingMockAPI(async mock => {
            const nav = SirenNav.create("http://localhost/api");
            const foo = nav.performAction("query", { term: "home", x: 5 });
            const r1 = await foo.asSiren();
            expect(r1).toMatchSnapshot();

            expect(mock.history.get).toHaveLength(2);
            expect(mock.history.get[1].method).toEqual("get");
            expect(mock.history.get[1].data).toEqual("?term=home&x=5");
        }),
    );
    it(
        "should perform a hypermedia action",
        usingMockAPI(async mock => {
            const nav = SirenNav.create("http://localhost/api");
            const foo = nav.performHyperAction<undefined>("query", {
                properties: undefined,
                links: [],
            });
            const r1 = await foo.asSiren();
            expect(r1).toMatchSnapshot();
        }),
    );
    it.skip(
        "should perform a POST action",
        usingMockAPI(async () => {
            const nav = SirenNav.create("http://localhost/api");
            const bar = nav.performAction("create", {});
            const r2 = await bar.asSiren();
            expect(r2).toMatchSnapshot();
        }),
    );
});
