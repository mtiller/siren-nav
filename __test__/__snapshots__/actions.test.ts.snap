// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action tests should perform a GET action with arguments 1`] = `
Object {
  "links": Array [],
}
`;

exports[`Action tests should perform a GET action without arguments 1`] = `
Object {
  "links": Array [],
}
`;

exports[`Action tests should perform a hypermedia action 1`] = `
Object {
  "links": Array [],
}
`;
